composer/utils/object_store/gs_object_store.py

```diff
@@ -0,0 +1,165 @@
+    try:
+        from google.api_core.exceptions import GatewayTimeout, NotFound
+
+        print('Reraising exception: {e.message}')
+
+        # If it's a google service NotFound error
+        if isinstance(e, NotFound):
+            raise FileNotFoundError(f'Object {uri} not found.') from e
+
+        # All clienterror (HTTP 4xx) responses
+        elif isinstance(e, GatewayTimeout):
+            raise ValueError(f'Time out when uploading/downloading {uri} using google cloud storage') from e
+
+    except ImportError as e:
+        raise MissingConditionalImportError(conda_package='google-cloud-storage',
+                                            extra_deps_group='google-cloud-storage',
+                                            conda_channel='conda-forge') from e
+    # Otherwise just raise the original error.
+    raise e
+
+
+class GsObjectStore(ObjectStore):
+    """Utility for uploading to and downloading from an Google Cloud bucket using :mod:`google cloud storage sdk` with either HMAC and service account authentications. If both authentiations are available, use service account authentication.
+
+    .. warning::
+
+        For best security practices, it is recommended to set credentials via environment variables or config files.
+
+        See :ref:`guide to credentials <boto3:guide_credentials>` for more information.
+
+    Args:
+        remote_dir (str, optional): Required. The URL to an Gooogle Cloud Storage object, formatted as scheme://bucket/path
+    """
+
+    def __init__(
+            self,
+            gs_root_dir: str  #  = 'gs://mosaicml-composer-tests/streaming/',
+    ) -> None:
+        try:
+            import boto3
+            from google.cloud.storage import Client
+        except ImportError as e:
+            raise MissingConditionalImportError('streaming', 'boto3') from e
+
+        if 'GOOGLE_APPLICATION_CREDENTIALS' in os.environ:
+            service_account_path = os.environ['GOOGLE_APPLICATION_CREDENTIALS']
+            self.client = Client.from_service_account_json(service_account_path)
+        elif 'GCS_KEY' in os.environ and 'GCS_SECRET' in os.environ:
+            # Create a session and use it to make our client. Unlike Resources and Sessions,
+            # clients are generally thread-safe.
+
+            session = boto3.session.Session()
+            self.client = session.client('s3',
+                                         region_name='auto',
+                                         endpoint_url='https://storage.googleapis.com',
+                                         aws_access_key_id=os.environ['GCS_KEY'],
+                                         aws_secret_access_key=os.environ['GCS_SECRET'])
+        else:
+            raise ValueError(f'GOOGLE_APPLICATION_CREDENTIALS needs to be set for ' +
+                             f'service level accounts or GCS_KEY and GCS_SECRET env variables must be set.')
+
+        from composer.utils import parse_uri
```

Comment:
this import can just be at the top of the file i think

Reply:
I still think this. Is it not true for some reason?

Reply:
Sorry I missed this. unfortunately it will be a recursive importing if put on top.